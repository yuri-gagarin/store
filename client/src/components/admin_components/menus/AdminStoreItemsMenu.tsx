import React, { useState, useEffect, useRef } from "react";
import { Dropdown, DropdownProps, Menu, MenuItemProps } from "semantic-ui-react";
// routing //
import { withRouter, RouteComponentProps, useRouteMatch } from "react-router-dom"
// css imports //
import "./css/adminStoreItemsMenu.css";
// actions and state //
import { AppAction, IGlobalAppState } from "../../../state/Store";
import { getAllStores } from "../stores/actions/APIstoreActions";
import { getAllStoreItems } from "../store_items/actions/APIStoreItemActions";
// helpers //
import { capitalizeString } from "../../helpers/displayHelpers";

interface Props extends RouteComponentProps {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
interface DProps {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>
}
type DropdownData = {
  key: string;
  text: string;
  value: string;
}
const StoreNameDropDown: React.FC<DProps> = ({ state, dispatch }): JSX.Element => {
  const [ dropdownState, setDropdownState ] = useState<DropdownData[]>();
  const { loadedStores } = state.storeState;

  const handleSearchChange = (e: React.SyntheticEvent, data: DropdownProps): void => {
    console.log(31);
    console.log(e)
    console.log(data.value);
    const queryOptions = {
      storeName: data.value as string
    }
    getAllStoreItems(dispatch, queryOptions)
  }
  useEffect(() => {
    getAllStores(dispatch)
  }, [])
  useEffect(() => {
    const dropdownData = loadedStores.map((store) => {
      return {
        key: store._id,
        text: capitalizeString(store.title),
        value: store.title
      }
    });
    setDropdownState(() => {
      return [ ...dropdownData ];
    })
  }, [loadedStores]);
  return (
    <Dropdown
      placeholder={"Filter by Store name"}
      selection
      onChange={handleSearchChange}
      options={dropdownState}
    />
  )
}
const AdminStoreItemsMenu: React.FC<Props> = ({ history, location, state, dispatch }): JSX.Element => {
  const [ scrolled, setScrolled ] = useState<boolean>(false);
  const [ activeItem, setActiveItem ] = useState<string>("view_all");
  const [ menuOpen, setMenuOpen ] = useState<boolean>(false);

  const match = useRouteMatch("/admin/home/store_items");
  const adminStoreItemMenuRef = useRef<HTMLDivElement>(document.createElement("div"));

  const handleItemClick = (e: React.MouseEvent, { name }: MenuItemProps): void => {
    setActiveItem(String(name));


    switch (name) {
      case "view_all": 
        history.push(match?.path + "/all");
        break;
      case "create":
        history.push(match?.path + "/create");
        dispatch({ type: "CLEAR_CURRENT_STORE_ITEM", payload: null });
        break;
      case "manage":
        history.push(match?.path + "/manage");
        break;
      default: history.push("/admin/home");
    }
  }
 
  const scrollListener = () => {
    if (window.scrollY > 1) {
      setScrolled(true);
    } else if (window.scrollY === 0) {
      setScrolled(false)
    } 
  };
  // lifecycle hooks //
  useEffect(() => {
    const currentURL = location.pathname;
    setTimeout(() => {
      setMenuOpen(true);
    }, 200);
    if (currentURL.match(/all/)) {
      setActiveItem("view_all");
    } else if (currentURL.match(/create/)) {
      setActiveItem("create");
    } else if (currentURL.match(/manage/)) {
      setActiveItem("manage");
    }
    // menu animation timeout //
    setTimeout(() => {
      setMenuOpen(true);
    }, 200);
  }, []);

  useEffect(() => {

    window.addEventListener("scroll", scrollListener, true);
    return () => window.removeEventListener("scroll", scrollListener);

  }, [adminStoreItemMenuRef]);

  return (
    <div className={ scrolled ? "adminStoreItemsMenuFixed menuScrolled" : "adminStoreItemsMenuFixed"} ref={adminStoreItemMenuRef}>
      <Menu tabular className={ menuOpen ? "adminStoreItemsMenu storeItemsMenuOpen" : "adminStoreItemsMenu" }>
        <StoreNameDropDown state={state} dispatch={dispatch} />
        <Menu.Item
          name='view_all'
          content="View All StoreItems"
          active={activeItem === 'view_all'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='create'
          content="Create New StoreItem"
          active={activeItem === 'create'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='manage'
          content="Manage StoreItems"
          active={activeItem === 'manage'}
          onClick={handleItemClick}
        />
      </Menu>
    </div>
  )
};

export default withRouter(AdminStoreItemsMenu);