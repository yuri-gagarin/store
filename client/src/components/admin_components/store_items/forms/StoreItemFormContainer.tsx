import React, { useContext, useState, useEffect } from "react";
import { Button, Grid, List } from "semantic-ui-react";
// css imports //
import "./css/storeItemFormContainer.css";
// additional components //
import StoreItemForm from "./StoreItemForm";
import StoreItemImgPreviewContainer from "../image_preview/StoreItemImgPreviewContainer";
import StoreItemImgUplForm from "./StoreItemImgUplForm";
import FormErrorComponent from "../../popups/FormErrorComponent";
import LoadingBar from "../../miscelaneous/LoadingBar";
// state //
import { Store } from "../../../../state/Store";
// client routing //
import { RouteComponentProps, withRouter } from "react-router-dom";
// api actions //
import { createStoreItem, editStoreItem } from "../actions/APIStoreItemActions";
import { getAllStores } from "../../stores/actions/APIstoreActions";
// ui actions //
import { openStoreItemForm, closeStoreItemForm } from "../actions/UIStoreItemActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// types 
import { AxiosError } from "axios";
import { StoreItemFormState, StoreItemData, StoreDropdownData } from "../type_definitions/storeItemTypes";

interface Props extends RouteComponentProps {

}
const StoreItemFormContainer: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentStoreItemData, storeItemFormOpen, error } = state.storeItemState;
  const { storeId, storeName, name, description, details, price, categories, createdAt, editedAt } = currentStoreItemData;
  // local component state //
  // const [ formOpen, setFormOpen ] = useState<boolean>(false);
  const [ newForm, setNewForm ] = useState<boolean>(true);
  const [ storeDropDownData, setStoreDropDownData ] = useState<StoreDropdownData[]>();
  
  // StoreItemForm toggle //
  const toggleForm = () => {
    storeItemFormOpen ? closeStoreItemForm(dispatch) : openStoreItemForm(dispatch);
  };
  // API call handlers CREATE - EDIT //
  const handleCreateStoreItem = ({ storeId, storeName, name, price, description, details, categories }: StoreItemFormState): void => {
    const storeItemData: StoreItemData = {
      storeId,
      storeName,
      name,
      description,
      details,
      price,
      images: currentStoreItemData.images,
      categories,
    };
    createStoreItem(storeItemData, dispatch)
      .then((_) => {
        // storeItem created //
        closeStoreItemForm(dispatch);
        // push to view Item
        // history.push("")
      })
      .catch((_) => {
        // handle error? show some modal? //
      });
  };
  const handleUpdateStoreItem = ({ storeId, storeName,name, description, details, price, categories }: StoreItemFormState): void => {
    const storeItemParams: StoreItemData = {
      storeId, storeName, name, details, description, price, images: currentStoreItemData.images, categories
    };
    editStoreItem(currentStoreItemData._id, storeItemParams, dispatch, state)
      .then((_) => {
        closeStoreItemForm(dispatch);
      })
      .catch((_) => {
        // handle error show modal ? //
      });
  };
  // lifecycle hooks //
  useEffect(() => {
    let mounted = true;
    if (mounted) {
      getAllStores(dispatch)
        .then((_) => {
          // handle success //
          const { loadedStores } = state.storeState;
          const storeDropdownData: StoreDropdownData[] = loadedStores.map((store) => {
            return {
              storeId: store._id,
              storeName: store.title
            }
          });
          setStoreDropDownData(storeDropdownData);
          checkSetValues(currentStoreItemData) ? setNewForm(false) : setNewForm(true);
        })
        .catch((_) => {
          // handle error //
        });
      }
    return () => { mounted = false };
  }, []);
  useEffect(() => {
    if (!storeItemFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ storeItemFormOpen ]);
  
  useEffect(() => {
    checkSetValues(currentStoreItemData) ? setNewForm(false) : setNewForm(true);
  }, [ currentStoreItemData]);
  // component return //
  return (
    <div id="storeItemFormContainer">
      { loading ? <LoadingBar /> : null }
      <FormErrorComponent error={error as AxiosError} />
      {
        !newForm ?
          <div id="storeItemFormContainerDetailsContainer">
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="storeItemFormContainerDetail">
                  <h3>Store Item name:</h3>
                  <p>{name}</p>
                </div>
                <div className="storeItemFormContainerDetail">
                  <h3>Store Item price:</h3>
                  <p>{price}</p>
                </div>
                <div className="storeItemFormContainerDetail">
                  <h3>Store Item description:</h3>
                  <p>{description}</p>
                </div>
                <div className="storeItemFormContainerDetail">
                  <h3>Store Item details:</h3>
                  <p>{details}</p>
                </div>
                <div className="storeItemFormContainerCategories">
                  <h3>Listed Store Item categories: </h3>
                  { categories.length === 0 ? "no categories listed" : null}
                  <List horizontal>
                    {
                      categories.map((category) => {
                        return (
                          <List.Item key={category}>
                            <List.Content>
                              <List.Header>{category}</List.Header>
                            </List.Content>
                          </List.Item>
                        )
                      })
                    }
                  </List>
                </div>
                <div className="storeItemFormContainerTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <StoreItemImgPreviewContainer state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <StoreItemImgUplForm />
          </div>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  id="storeItemFormToggleBtn" 
            onClick={toggleForm} 
            content={ !storeItemFormOpen ? "Open Form" : "Close Form"}
          />
          {
            storeItemFormOpen ? 
              <StoreItemForm 
                storeId={storeId}
                storeName={storeName}
                name={name}
                price={price}
                details={details}
                description={description}
                categories={categories}
                newForm={newForm}
                activeStores={storeDropDownData ? storeDropDownData : []}                
                handleCreateStoreItem={handleCreateStoreItem}
                handleUpdateStoreItem={handleUpdateStoreItem}
              /> 
            : null
          }
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};
// export for render tests //
export { StoreItemFormContainer }
// default export //
export default withRouter(StoreItemFormContainer);

