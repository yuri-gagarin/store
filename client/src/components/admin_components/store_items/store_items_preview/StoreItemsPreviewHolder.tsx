import React, { useContext, useEffect, useRef, useState } from "react";
import { Grid, Item } from "semantic-ui-react";
// routing //
import { withRouter, RouteComponentProps } from "react-router-dom";
// css imports //
import "./css/storeItemsPreviewHolder.css";
// additional components //
import StoreItemPreview from "./StoreItemPreview";
import StoreItemsControls from "./StoreItemsControls";
import PopularStoreItemsHolder from "./popular_store_items/PopularStoreItemsHolder";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
import ErrorScreen from "../../miscelaneous/ErrorScreen";
// types and interfaces //
import { Store } from "../../../../state/Store";
// api actions //
import { getAllStoreItems } from "../actions/APIStoreItemActions";

interface Props extends RouteComponentProps {
  
}

const StoreItemsPreviewHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, loadedStoreItems, error } = state.storeItemState;
  // local state //
  const [ newDataLoaded, setNewDataLoaded ] = useState<boolean>(false);
  const storeItemsRef = useRef(loadedStoreItems);

  // lifecycle hooks //
  useEffect(() => {
    let componentLoaded = true;
    if (componentLoaded) {
      getAllStoreItems(dispatch)
        .then((_) => {
          setNewDataLoaded(true);
        })
        .catch((_) => {
          // handle error show modal ? //
          setNewDataLoaded(false);
        });
    }
    return () => { componentLoaded = false };
  }, [ dispatch ]);

  useEffect(() => {
    if(storeItemsRef.current !== loadedStoreItems && !loading && !error) {
      setNewDataLoaded(true);
    } 
  }, [ storeItemsRef.current, loadedStoreItems, loading, error ]);
  // component return //
  return (
    newDataLoaded ?
    <Grid stackable padded columns={2}>
      <Grid.Row>
      <Grid.Column computer={10} tablet={8} mobile={16}>
        <Item.Group>
          {
            loadedStoreItems.map((storeItem) => {
              return (
                <StoreItemPreview 
                  key={storeItem._id}
                  storeItem={storeItem}
                />
              );
            })
          }
        </Item.Group>
      </Grid.Column>
      <Grid.Column computer={6} tablet={8} mobile={16}>
        <StoreItemsControls state={state} dispatch={dispatch} />
        <PopularStoreItemsHolder popularStoreItems={loadedStoreItems}/>
      </Grid.Column>

      </Grid.Row>
      
    </Grid>
    : 
    (
      error ? <ErrorScreen lastRequest={ () => getAllStoreItems(dispatch) } /> : <LoadingScreen />
    )
  );
};
// test export without the router //
export { StoreItemsPreviewHolder };
// default export //
export default withRouter(StoreItemsPreviewHolder);