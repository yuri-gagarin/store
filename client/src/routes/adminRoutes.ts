export const AdminProductRoutes = {
  HOME_ROUTE: "/admin/home/my_products",
  VIEW_ALL_ROUTE: "/admin/home/my_products/view_all",
  CREATE_ROUTE: "/admin/home/my_products/create",
  MANAGE_ROUTE: "/admin/home/my_products/manage",
  SHOW_ROUTE: "/admin/home/my_products/show",
  VIEW_SORTED_ROUTE: "/admin/home/my_products/view_sorted"
};

export const AdminServiceRoutes = {
  ADMIN_SERVICES_HOME_ROUTE: "/admin/home/my_services",
  ADMIN_SERVICES_VIEW_ALL_ROUTE: "/admin/home/my_services/view_all",
  ADMIN_SERVICES_CREATE_ROUTE: "/admin/home/my_services/create",
  ADMIN_SERVICES_MANAGE_ROUTE: "/admin/home/my_services/manage",
  ADMIN_SERVICES_SHOW_ROUTE: "/admin/home/my_services/show",
  ADMIN_SERVICES_VIEW_SORTED_ROUTE: "/admin/home/my_services/view_sorted"
};

export const AdminStoreRoutes = {
  ADMIN_STORES_HOME_ROUTE: "/admin/home/my_stores",
  ADMIN_STORES_CREATE_ROUTE: "/admin/home/my_stores/create",
  ADMIN_STORES_MANAGE_ROUTE: "/admin/home/my_stores/manage",
  ADMIN_STORES_SHOW_ROUTE: "/admin/home/my_stores/show",
  ADMIN_STORES_VIEW_ALL_ROUTE: "/admin/home/my_stores/view_all"
};

export const AdminStoreItemRoutes = {
  HOME_ROUTE: "/admin/home/my_store_items",
  CREATE_ROUTE: "/admin/home/my_store_items/create",
  MANAGE_ROUTE: "/admin/home/my_store_items/manage",
  SHOW_ROUTE: "/admin/home/my_store_items/show",
  VIEW_ALL_ROUTE: "/admin/home/my_store_items/view_all",
  VIEW_SORTED_ROUTE: "/admin/home/my_store_items/view_sorted"
};

