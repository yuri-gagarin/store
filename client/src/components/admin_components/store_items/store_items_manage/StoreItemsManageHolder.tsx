import React, { useContext, useEffect, useRef, useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import StoreItemFormHolder from "../forms/StoreItemFormHolder";
import StoreItemCard from "./StoreItemCard";
import StoreItemControls from "../store_items_preview/StoreItemsControls";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
import ErrorScreen from "../../miscelaneous/ErrorScreen";
// actions and state //
import { getAllStoreItems } from "../actions/APIStoreItemActions";
import { Store } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps, useRouteMatch, Route } from "react-router-dom";
import { AdminStoreItemRoutes } from "../../../../routes/adminRoutes";

interface Props extends RouteComponentProps {
};

const StoreItemsManageHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, loadedStoreItems, error } = state.storeItemState;
  // local state //
  const [ newDataLoaded, setNewDataLoaded ] = useState<boolean>(false);
  const storeItemsRef = useRef(loadedStoreItems);
  // client routing //
  const match = useRouteMatch(AdminStoreItemRoutes.MANAGE_ROUTE);

  const handleBack = () => {
    history.goBack();
  };
  // lifecycle hooks //
  useEffect(() => {
    let isMounted = true;
    if (isMounted) {
      getAllStoreItems(dispatch)
        .then((_) => {
          setNewDataLoaded(true);
        })
        .catch((_) => {
          setNewDataLoaded(false);
        });
    }
    return () => { isMounted = false };
  }, [ dispatch ]); 

  useEffect(() => {
    if (storeItemsRef.current != loadedStoreItems && !loading && !error) {
      setNewDataLoaded(true);
    }
  }, [ storeItemsRef.current, loadedStoreItems, loading, error ]);

  return (
    newDataLoaded ? 
    <Grid padded stackable columns={2}>
      <Route path={AdminStoreItemRoutes.EDIT_ROUTE}> 
        <Grid.Row>
          <Grid.Column computer={12} tablet={6} mobile={16}>
            <h3>Editing Store Item: { state.storeItemState.currentStoreItemData.name }</h3>
            <Button
              id="adminStoreItemsManageBackBtn"
              inverted 
              color="green" 
              content="Back" 
              onClick={handleBack}>
            </Button>
          </Grid.Column>
        </Grid.Row>
        <StoreItemFormHolder />
      </Route>
      <Route exact path={match?.url}>
        <Grid.Row>
          <Grid.Column computer={12} tablet={8} mobile={16}>
          {
            loadedStoreItems.map((storeItem) => {
              return (
                <StoreItemCard 
                  key={storeItem._id}
                  storeItem={storeItem}
                  imageCount={storeItem.images.length}
                  state={state}
                  dispatch={dispatch}
                />
              );
            })
          }
          </Grid.Column>
          <Grid.Column computer={4} tablet={8} mobile={16}>
            <StoreItemControls state={state} dispatch={dispatch} />
          </Grid.Column>
        </Grid.Row>
      </Route>
    </Grid>
    :
    (
      error ? <ErrorScreen lastRequest={ () => getAllStoreItems(dispatch) } /> : <LoadingScreen />
    )
  );
}
// for tests //
export { StoreItemsManageHolder };
// 
export default withRouter(StoreItemsManageHolder);