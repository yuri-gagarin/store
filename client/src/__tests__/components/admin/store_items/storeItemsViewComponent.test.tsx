import React, { ReactElement } from "react";
import { Menu } from "semantic-ui-react";
import { mount, ReactWrapper } from "enzyme";
// component imports //
import StoreItemsViewComponent from "../../../../components/admin_components/store_items/StoreItemsView";
import AdminStoreItemsMenu from "../../../../components/admin_components/menus/AdminStoreItemsMenu";
import { StoreItemsPreviewHolder } from "../../../../components/admin_components/store_items/store_items_preview/StoreItemsPreviewHolder";
import { StoreItemFormContainer } from "../../../../components/admin_components/store_items/forms/StoreItemFormContainer";
// additional dependencies //
import  { BrowserRouter as Router, Switch, MemoryRouter } from "react-router-dom";
import { StoreItemsManageHolder } from "../../../../components/admin_components/store_items/store_items_manage/StoreItemsManageHolder";
import { AdminStoreItemRoutes } from "../../../../routes/adminRoutes";

describe("StoreItemView Component test", () => {
  let wrapper: ReactWrapper;
  beforeAll(() => {
    wrapper = mount<{}, typeof Router>(
      <MemoryRouter initialEntries={[AdminStoreItemRoutes.HOME_ROUTE]} keyLength={0}>
        <StoreItemsViewComponent />
      </MemoryRouter>
    );
  });
  describe("StoreItemView Component render test", () => {
    it("Should Properly render StoreItemsView", () => {
      expect(wrapper.find(StoreItemsViewComponent)).toMatchSnapshot();
    });
    it("Should render Admin StoreItem Menu", () => {
      expect(wrapper.find(AdminStoreItemsMenu)).toHaveLength(1);
    });
    it("Should render conditional routes", () => {
      expect(wrapper.find("Switch")).toHaveLength(1);
    });
    it("Should have Render three children route componets", () => {
      const switchComponent = wrapper.find(Switch);
      expect(switchComponent.props().children).toHaveLength(3);
    });
    it("Should have proper client routes and conditionally render correct Components", () => {
      const switchComponent = wrapper.find(Switch);
      type RouteMap = {
        [key: string]: string
      }
      let map: RouteMap = {};

      const routeComponents = switchComponent.props().children as ReactElement[];
      for (const component of routeComponents) {
        if (component.props.children[1].type.WrappedComponent) {
          // to check if a component is wrapped in a WithRouter() function //
          map[component.props.path as string] = component.props.children[1].type.WrappedComponent;
        } else {
          map[component.props.path as string] = component.props.children[1].type;
        }
      }
      expect(map[AdminStoreItemRoutes.VIEW_ALL_ROUTE]).toBe(StoreItemsPreviewHolder);
      expect(map[AdminStoreItemRoutes.CREATE_ROUTE]).toBe(StoreItemFormContainer);
      expect(map[AdminStoreItemRoutes.MANAGE_ROUTE]).toBe(StoreItemsManageHolder);
    });
  });
  
  describe("StoreItemView Component button actions", () => {
    describe("'click' Event on 'View All' Link", () => {
      it("Should render 'StoreItemsPreviewHolder' component", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = wrapper.find(AdminStoreItemsMenu).find(Menu.Item).at(0);
        viewAllLink.simulate("click");
        wrapper.update();
        expect(wrapper.find(StoreItemsPreviewHolder).length).toEqual(1);
      });
      it("Should NOT render 'StoreItemFormHolder' component", () => {
        expect(wrapper.find(StoreItemFormContainer).length).toEqual(0);
      });
      it("Should NOT render 'StoreItemsManageHolder' compnent", () => {
        expect(wrapper.find(StoreItemsManageHolder).length).toEqual(0);
      });
    });
    describe("'click' Event on 'Create Store Item' Link", () => {
      it("Should render 'StoreItemsFormHolder' component", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = wrapper.find(AdminStoreItemsMenu).find(Menu.Item).at(1);
        viewAllLink.simulate("click");
        wrapper.update();
        expect(wrapper.find(StoreItemFormContainer).length).toEqual(1);
      });
      it("Should NOT render 'StoreItemsPreviewHolder' component", () => {
        expect(wrapper.find(StoreItemsPreviewHolder).length).toEqual(0);
      });
      it("Should NOT render 'StoreItemsManageHolder' component", () => {
        expect(wrapper.find(StoreItemsManageHolder).length).toEqual(0);
      });
    });
    describe("'click' Event on 'Manage Store Items' Link", () => {
      it("Should render 'StoreItemsManage' component", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = wrapper.find(AdminStoreItemsMenu).find(Menu.Item).at(2);
        viewAllLink.simulate("click");
        wrapper.update();
        expect(wrapper.find(StoreItemsManageHolder).length).toEqual(1);
      });
      it("Should NOT redner 'StoreItemsFormHolder' component", () => {
        expect(wrapper.find(StoreItemFormContainer).length).toEqual(0);
      });
      it("Should NOT redner 'StoreItemsPreviewHolder' component", () => {
        expect(wrapper.find(StoreItemsPreviewHolder).length).toEqual(0);
      });
    });
    
  });
  
});