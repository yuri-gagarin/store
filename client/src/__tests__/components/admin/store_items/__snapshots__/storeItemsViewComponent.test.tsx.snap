// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreItemView Component test StoreItemView Component render test Should Properly render StoreItemsView 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/admin/home/store_items",
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "vy8a1q",
            "pathname": "/admin/home/store_items",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "vy8a1q",
          "pathname": "/admin/home/store_items",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <StoreItemsView>
      <div
        id="adminStoreItemsViewHolder"
      >
        <withRouter(AdminStoreItemsMenu)
          dispatch={[Function]}
          state={
            Object {
              "bonusVideoState": Object {
                "currentBonusVideoData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "vimeoURL": "",
                  "youTubeURL": "",
                },
                "error": null,
                "loadedBonusVideos": Array [],
                "loading": false,
                "responseMsg": "",
              },
              "productState": Object {
                "currentProductData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "details": "",
                  "images": Array [],
                  "name": "",
                  "price": "",
                },
                "error": null,
                "loadedProducts": Array [],
                "loading": false,
                "responseMsg": "",
              },
              "serviceState": Object {
                "currentServiceData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "images": Array [],
                  "name": "",
                  "price": "",
                },
                "error": null,
                "loadedServices": Array [],
                "loading": false,
                "responseMsg": "",
              },
              "storeItemState": Object {
                "currentStoreItemData": Object {
                  "_id": "",
                  "categories": Array [],
                  "createdAt": "",
                  "description": "",
                  "details": "",
                  "images": Array [],
                  "name": "",
                  "price": "",
                  "storeId": "",
                  "storeName": "",
                },
                "error": null,
                "loadedStoreItems": Array [],
                "loading": false,
                "numberOfItems": 0,
                "responseMsg": "",
              },
              "storeState": Object {
                "currentStoreData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "images": Array [],
                  "title": "",
                },
                "error": null,
                "loadedStores": Array [],
                "loading": false,
                "responseMsg": "",
              },
            }
          }
        >
          <AdminStoreItemsMenu
            dispatch={[Function]}
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "key": "vy8a1q",
                    "pathname": "/admin/home/store_items",
                    "search": "",
                    "state": undefined,
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "vy8a1q",
                  "pathname": "/admin/home/store_items",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "key": "vy8a1q",
                "pathname": "/admin/home/store_items",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": false,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            state={
              Object {
                "bonusVideoState": Object {
                  "currentBonusVideoData": Object {
                    "_id": "",
                    "createdAt": "",
                    "description": "",
                    "vimeoURL": "",
                    "youTubeURL": "",
                  },
                  "error": null,
                  "loadedBonusVideos": Array [],
                  "loading": false,
                  "responseMsg": "",
                },
                "productState": Object {
                  "currentProductData": Object {
                    "_id": "",
                    "createdAt": "",
                    "description": "",
                    "details": "",
                    "images": Array [],
                    "name": "",
                    "price": "",
                  },
                  "error": null,
                  "loadedProducts": Array [],
                  "loading": false,
                  "responseMsg": "",
                },
                "serviceState": Object {
                  "currentServiceData": Object {
                    "_id": "",
                    "createdAt": "",
                    "description": "",
                    "images": Array [],
                    "name": "",
                    "price": "",
                  },
                  "error": null,
                  "loadedServices": Array [],
                  "loading": false,
                  "responseMsg": "",
                },
                "storeItemState": Object {
                  "currentStoreItemData": Object {
                    "_id": "",
                    "categories": Array [],
                    "createdAt": "",
                    "description": "",
                    "details": "",
                    "images": Array [],
                    "name": "",
                    "price": "",
                    "storeId": "",
                    "storeName": "",
                  },
                  "error": null,
                  "loadedStoreItems": Array [],
                  "loading": false,
                  "numberOfItems": 0,
                  "responseMsg": "",
                },
                "storeState": Object {
                  "currentStoreData": Object {
                    "_id": "",
                    "createdAt": "",
                    "description": "",
                    "images": Array [],
                    "title": "",
                  },
                  "error": null,
                  "loadedStores": Array [],
                  "loading": false,
                  "responseMsg": "",
                },
              }
            }
          >
            <div
              className="adminStoreItemsMenuFixed"
            >
              <Menu
                className="adminStoreItemsMenu"
                tabular={true}
              >
                <div
                  className="ui tabular adminStoreItemsMenu menu"
                >
                  <MenuItem
                    active={true}
                    content="View All StoreItems"
                    name="view_all"
                    onClick={[Function]}
                  >
                    <a
                      className="active item"
                      onClick={[Function]}
                    >
                      View All StoreItems
                    </a>
                  </MenuItem>
                  <MenuItem
                    active={false}
                    content="Create New StoreItem"
                    name="create"
                    onClick={[Function]}
                  >
                    <a
                      className="item"
                      onClick={[Function]}
                    >
                      Create New StoreItem
                    </a>
                  </MenuItem>
                  <MenuItem
                    active={false}
                    content="Manage StoreItems"
                    name="manage"
                    onClick={[Function]}
                  >
                    <a
                      className="item"
                      onClick={[Function]}
                    >
                      Manage StoreItems
                    </a>
                  </MenuItem>
                </div>
              </Menu>
            </div>
          </AdminStoreItemsMenu>
        </withRouter(AdminStoreItemsMenu)>
        <Switch />
      </div>
    </StoreItemsView>
  </Router>
</MemoryRouter>
`;
